java/org/apache/jasper/compiler/Generator.java

```diff
@@ -2095,6 +2104,20 @@ public void visit(Node.JspElement n) throws JasperException {
         public void visit(Node.TemplateText n) throws JasperException {
 
             String text = n.getText();
+            // If the flag is active, attempt to minimize the frequency of
+            // regex operations.
+            if ((ctxt!=null) &&
+                ctxt.getOptions().getJSPWhiteSpaceTrimFlag() &&
+                text.contains("\n")) {
+                // Ensure there are no <pre> or </pre> tags embedded in this
+                // text - if there are, we want to NOT modify the whitespace.
+                Matcher preMatcher = PRE_TAG_PATTERN.matcher(text);
```

Comment:
Are we guaranteed to have well-formed tags here (i.e. open/close) or could this split across tags?  What are examples of text we might expect here?